src/Streamly/Coreutils/Cp.hs

```diff
@@ -0,0 +1,281 @@
+
+    identify :: FilePath -> FilePath -> IO (Either FilePath FilePath)
+    identify dir path = do
+       let actualPath = append baseDir $ append dir path
+       isFile <- doesFileExist actualPath
+       if isFile == True
+       then
+           return $ Right $ append dir path     -- RELATIVE File
+       else
+           return $ Left $ append dir path      -- RELATIVE Dir
+
+    contents :: FilePath -> [FilePath]
+    contents dir = unsafePerformIO $ listDirectory $ append baseDir dir
+
+
+
+-- | Recursively copies the contents of the source directory
+-- to the destination directory.
+-- Arguments are of type @FilePath@.
+--
+-- @since 0.1.0.0
+copyDirToDir
+    :: IsStream t
+    => CpOptions
+    -> FilePath
+    -> FilePath
+    -> t IO ()
+copyDirToDir opt src dest = do
+    if force opt == True
+    then
+        S.mapM handleForce
+        $ traverseDir src
+    else
+        S.mapM (\_ -> return ())
+        $ S.takeWhileM handleNonForce
+        $ traverseDir src
+
+    where
+
+    handleForce :: Either FilePath FilePath -> IO ()
+    handleForce (Left dir) = createDirectoryIfMissing True $ append dest dir
+    handleForce (Right file) =
+        ( copy opt (append src file) (append dest file)
+        >> return ()
+        )
+
+    handleNonForce :: Either FilePath FilePath -> IO Bool
+    handleNonForce (Left dir) =
+        ( createDirectoryIfMissing True (append dest dir)
+        >> return True
+        )
+    handleNonForce (Right file) =
+        copy opt (append src file) (append dest file)
+
+
+-- | Copies the source file to the
+-- destination path with a possible rename.
+--
+-- @since 0.1.0.0
+cpFileWithRename
+    :: CpOptions
+    -> SomeBase File
+    -> SomeBase File
+    -> IO Bool
+cpFileWithRename opt src dst = do
+    let srcFp = removeTrailingSlash $ someFileToFP src
+    let dstFp = removeTrailingSlash $ someFileToFP dst
+    copy opt srcFp dstFp
+
+
+-- | Copies the source file preserving the source file name
+-- in the destination directory.
+-- Returns @True@ in the @IO@ monad if copying was successful,
+-- @False@ otherwise.
+--
+-- @since 0.1.0.0
+cpFile
+    :: CpOptions
+    -> SomeBase File
+    -> SomeBase Dir
+    -> IO Bool
+cpFile option src dest = do
+    copy option srcFp destFp
+
+    where
+
+    srcFp = removeTrailingSlash $ someFileToFP src
+    dirFp = someDirToFP dest
+    destFp = append dirFp fileName
+    fileName = extractFileName src
+
+    -- | To extract the file part of a @SomeBase File@ path
+    extractFileName :: SomeBase File -> FilePath
+    extractFileName srcFile =
+        case srcFile of
+            Abs fpath -> fromRelFile $ filename fpath
+            Rel fpath -> fromRelFile $ filename fpath
+
+
+-- | Copies a stream of files to the destination directory.
+-- The stream can be drained with the appropriate combinator
+-- to achieve concurrency.
+-- Fails if the destination directory does not exist.
+--
+-- @since 0.1.0.0
+cpFiles
```

Comment:
What is `SomeBase`? Why not use `Path`?

Reply:
http://hackage.haskell.org/package/path-0.8.0/docs/Path-Posix.html#t:SomeBase